fix: added suitable filtering mechanism to link field

codescientist703 opened the pull request:
In the link field search widget, there is no sorting mechanism in place so sometimes user may not get the relevant query first in the search result. 
This PR introduces a sorting mechanism that does the following thing - 
1. All the search result which has user query as a prefix will come first and be sorted in lexicographical order.
2. All search result which contains user query will come last and sorted in lexicographical order.
This will ensure that users always get the relevant results first.
Before:
![Screenshot from 2021-05-20 19-18-37](https://user-images.githubusercontent.com/44427574/118990239-5496b180-b9a0-11eb-9a86-fdf50ddc6862.png)
After:
![Screenshot from 2021-05-20 19-19-03](https://user-images.githubusercontent.com/44427574/118990293-5eb8b000-b9a0-11eb-9d59-6db20cb13589.png)

closes #13216

gavindsouza commented on frappe/desk/search.py:
```diff
@@ -171,6 +171,7 @@ def search_widget(doctype, txt, query=None, searchfield=None, start=0,
 
 			if doctype in UNTRANSLATED_DOCTYPES:
 				values = tuple([v for v in list(values) if re.search(re.escape(txt)+".*", (_(v.name) if as_dict else _(v[0])), re.IGNORECASE)])
+				values = tuple(sorted(values, key= lambda x: ((_(x.name) if as_dict else _(x[0])).lower().startswith(txt.lower()) == False, (_(x.name) if as_dict else _(x[0])))))
```
Too convoluted to fit in a single line, and not readable. Multiple statements in a single line are not preferred ([PEP8](https://www.python.org/dev/peps/pep-0008/) [E701](https://www.flake8rules.com/rules/E701.html)). 

Also, try handing operations via L173 and 174 during a single iteration over the values. The existing logic can be simplified too.

codescientist703 replied:
okay, thanks for the review. I will do the necessary changes.

gavindsouza replied:
Maybe I wasn't clear enough earlier. Separate the list comprehension, use an appropriately named function instead of the lambda function. Even though it's styled across multiple lines, it's still a single statement expression, which is still slightly convoluted.

tldr; it's not easy to understand what's happening at first glance. Maybe add a comment that says 'why' and 'what' are you doing this for

codescientist703 replied:
@gavindsouza , sorry I thought that introducing a new function would be unnecessary that's why I used the lambda function. I have now done the changes as you asked.

codescientist703 commented:
@gavindsouza , I have done the changes as you asked.

gavindsouza commented:
Add explanations for your logic @codescientist703 

The change doesn't seem to be equivalent except for the sorting order. In the case of Tree doctypes, suggestions should include the child documents too. This worked before your change.

codescientist703 commented:
hey @gavindsouza , I have reverted back to use the previous filtering logic so my change is only adding the sorting logic which does not alter anything. I have also formatted the previous filtering logic code and added more explanations to my sorting logic.

gavindsouza commented:
![Screenshot 2021-06-08 at 5 53 45 PM](https://user-images.githubusercontent.com/36654812/121184218-7f2ea880-c882-11eb-9ecf-68f5c71b0ac9.png)

Not sure if this works

codescientist703 commented:
hey @gavindsouza , I am having some difficulty in creating tree doctype for testing purposes. I have created a doctype with a tree structure checkbox ticked. I was wondering  how can I make other doctypes as children of the doctype marked as Tree. Like in your case, children like New York, Washington have a parent as USA doctype.
PS - I finally figured it out by installing and testing on erpnext itself.

codescientist703 commented:
hey @gavindsouza , I found that the tree doctype issue you mentioned is a different issue altogether than the one I am trying to solve. I have pretty much fixed it by sorting but now wondering if this PR  should I remove the fix that I have done till now as it solves an entirely different problem?

codescientist703 committed:
fix: add sorting for paren tree doctype to appear above child

codescientist703 commented:
hey @gavindsouza , Since the original issue was that parent tree doctype was appearing below child doctype I have decided to fix that change only in this PR. I will add the other sorting and filtering change in another PR. Requesting to review the change.

gavindsouza commented:
@codescientist703 The issue is that the sorting order doesn't work based on the closest match. It's not supposed to be parent first or child first, rather whatever record is most likely to match user input first.

codescientist703 commented:
okay, I will try to incorporate the previous logic for the best input match.

codescientist703 committed:
fix: added sorting and filtering logic for relevant input results

codescientist703 commented:
hey @gavindsouza , I have merged my previous logic of sorting and filtering. I have also tested it on both erpnext and frappe framework link field and it's working seamlessly. It will now bring the best input match results.

codescientist703 commented:
hey @gavindsouza , Requesting review!

gavindsouza reviewed (commented):
This changes the suggestions given for a Link field pointing to a Tree DocType. Check out the Customer DocType in ERPNext for instance.

**Current Develop:**
![Screenshot 2021-07-05 at 11 02 50 AM](https://user-images.githubusercontent.com/36654812/124422316-b8264400-dd80-11eb-86ab-4a0c0c1e14ac.png)




**This Branch:**
![Screenshot 2021-07-05 at 11 03 41 AM](https://user-images.githubusercontent.com/36654812/124422301-b5c3ea00-dd80-11eb-8654-8a78a3a3e550.png)

Make sure the child rows aren't filtered out so we can take this further.

codescientist703 committed:
fix: child results should appear for parent search query

codescientist703 commented:
@gavindsouza , I have done the changes as asked.

gavindsouza committed:
refactor(search): Improvements in search_widget, search_link APIs

* Minor perf enhancements
* Renamed sorting_comparator to relevance_sorter

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into link-field-fix

gavindsouza commented:
LGTM. Made some minor changes. Can you add a test for this change so we can get this merged? @codescientist703

codescientist703 commented:
hey @gavindsouza , I have added the test for it.

gavindsouza commented on frappe/tests/test_search.py:
```diff
@@ -13,6 +13,11 @@ def test_search_field_sanitizer(self):
 		self.assertTrue('User' in result['value'])
 
+		# check if obtained result is sorted or not
+		search_link('DocType', '', query=None, filters=None, page_length=20, searchfield='name')
```
This should be added as a separate test, perhaps as `test_search_field_order` as you want to test the order...Also, mention a key instead of `''`. At it's current state, this test doesn't do much.

Even the coverage reduced after adding this 😅 


----

btw, the sider recommendation looks good. Incorporate it too?

codescientist703 replied:
If I put any key besides ' ', then I would have to use the `relevance_sorter` method for sorting and then comparing the result as the current method of sorting is just plain sorting. I didn't hardcode it as new doctypes would be kept on adding. So would that be fine @gavindsouza ?

codescientist703 replied:
yes, will incorporate that too,

codescientist703 commented:
@gavindsouza , I have added the test for `relevance_sorter` method so that way we need not test it on hardcoded doctype values and test coverage would also increase.

codescientist703 committed:
fix: fixed sider recommendation

codescientist703 commented:
@gavindsouza , I have added the test case which creates tree doctype

codescientist703 committed:
test: added test for order of link

gavindsouza reviewed (commented):
Test looks ok. However, `setUp` and `tearDown` methods are run before and after each test. Might need to make this conditional.

gavindsouza committed:
perf(test_search): Setup and teardown tests only if needed

codescientist703 commented:
> Test looks ok. However, `setUp` and `tearDown` methods are run before and after each test. Might need to make this conditional.

@gavindsouza , I have done that

gavindsouza reviewed (approved):
Breaking test is unrelated. LGTM

gavindsouza commented:
@Mergifyio backport version-13-hotfix
